pytensor/tensor/elemwise.py

```diff
@@ -818,18 +816,41 @@ def perform(self, node, inputs, output_storage):
             else:
                 storage[0] = variable
 
-    def infer_shape(self, fgraph, node, i_shapes) -> List[Tuple[TensorVariable, ...]]:
-        if len(node.outputs) > 1:
-            from pytensor.tensor.exceptions import ShapeError
-
-            raise ShapeError(
-                "Multiple outputs are not supported by the default `Elemwise.infer_shape`"
-            )
+    @staticmethod
+    def _check_runtime_broadcast(node, inputs):
+        for dims_and_bcast in zip(
+            *[
+                zip(input.shape, sinput.type.broadcastable)
+                for input, sinput in zip(inputs, node.inputs)
+            ]
+        ):
+            if any(d != 1 for d, _ in dims_and_bcast) and (1, False) in dims_and_bcast:
+                raise ValueError(
+                    "Runtime broadcasting not allowed. "
+                    "At least one input has a distinct dimension length of 1, but was not marked as broadcastable.\n"
+                    "If broadcasting was intended, use `specify_broadcastable` on the relevant input."
+                )
+                raise ValueError
 
-        out_shape = pytensor.tensor.broadcast_shape(*i_shapes, arrays_are_shapes=True)
+    def infer_shape(self, fgraph, node, i_shapes) -> List[Tuple[TensorVariable, ...]]:
```

Comment:
I didn't want to introduce checks or comparison between shapes, which that function does. This allows it to return a more optimized graph like Theano used to by assuming no invalid shapes were provided

The question then is whether we want to refactor that helper to do the same when `arrays_are_shapes=False”l`?

Reply:
I think the `make_node` is correct insofar as it uses static shape and it's not possible to have `broadcastable=False and shape=1` 

That one still requires some thinking and would be tackled in a separate PR.

Reply:
> I didn't want to introduce checks or comparison between shapes, which that function does. This allows it to return a more optimized graph like Theano used to by assuming no invalid shapes were provided

So we allow undefined behavior in the shapes *and* in rewrites? I'm not sure I see that much downside with having that check here...

But at least I think we shouldn't have this logic in both places. Maybe the function should have a flag if it should return shape with or without checks?

Reply:
I am thinking we should add a `config.assume_shapes_correct` flag (default to True) to toggle that behavior in both shape_inference and rewrites that can return simplified cases.

Reply:
Actually that helper works differently in that it expects either shapes or arrays, but here we are combining information from both shapes and arrays so it would require some refactoring. We don't want to simply pass `node.inputs` since `infer_shape` wants us to return a graph with `ishapes`.

I don't know if that is the right place to implement this logic since it is a user facing function. WDYT?

Reply:
Okay I reverted to using the helper. Thinks are a bit weird in shape compilation because it will just use the static type shape of the node if that's available. Because the Elemwise make_node assumes valid shapes, the check introduced by `infer_shape` is only triggered when all dims are `None`.

Reply:
Not much we can do about that then I think without a major rewrite of the shape handling...